Allow for optional can permission method for content review

satrun77 opened the pull request:
This PR allows for custom permission check for who can add a review of page content. If the permission method is not implemented, it will fall back to the previous check that uses the `canEdit` method.

The optional method `canReviewContent` can be added to any DataObject (ie. Page) to allow for a custom permission check to be applied. For example, a CMS user with read-only access to `Pages` section but can review page content and add comments. 

The unit tests ran successfully after this change with the exception of one test that was using an incorrect mocked return type. `ContentReviewCMSExtension::findRecord` [line 110](https://github.com/silverstripe/silverstripe-contentreview/blob/c5f2b44895338b16af6dababf6f1fe1009f469b4/src/Extensions/ContentReviewCMSExtension.php#L110)

The return type is `SiteTree` the unit test mock set return to `Member`

dhensby commented on src/Extensions/ContentReviewCMSExtension.php:
```diff
@@ -164,4 +165,18 @@ protected function getSchemaResponse($schemaID, $form = null, ValidationResult $
+        $isReviewer = $page->canView($user) && $page->hasMethod('canBeReviewedBy') && $page->canBeReviewedBy($user);
+        // Whether or not the user is allowed to review the content of the page
+        // Fallback to canEdit as it the original implementation
+        $canEdit = $page->hasMethod('canReviewContent')? $page->canReviewContent($user) : $page->canEdit();
```
```suggestion
        $canEdit = $page->hasMethod('canReviewContent') ? $page->canReviewContent($user) : $page->canEdit();
```

dhensby commented on src/Forms/ReviewContentHandler.php:
```diff
@@ -120,12 +120,18 @@ public function submitReview($record, $data)
+        $isReviewer = $record->canView($user) && $record->hasMethod('canBeReviewedBy') && $record->canBeReviewedBy($user);
+        // Whether or not a user is allowed to review the content of the record.
+        // Fallback to canEdit as it the original implementation
+        $canEdit = $record->hasMethod('canReviewContent')? $record->canReviewContent($user) : $record->canEdit();
```
```suggestion
        $canEdit = $record->hasMethod('canReviewContent') ? $record->canReviewContent($user) : $record->canEdit();
```

dhensby reviewed (dismissed):
There's some linting issues (including a line-length problem).

It would also be good to appease the code coverage and write a test that covers the missing code-paths.

emteknetnz commented on src/Forms/ReviewContentHandler.php:
```diff
@@ -120,12 +120,19 @@ public function submitReview($record, $data)
+
+        $user = Security::getCurrentUser();
+        // Whether or not the user is a reviewer. User must be allowed to view the
+        $isReviewer = $record->canView($user) &&
```
`$isReviewer` + `$canEdit` is code-duplication with the new `canUseReviewContent()` method above in this PR.  Is there a way that we can reduce this to a single-occurance?

satrun77 replied:
Can create a trait that contains the similar lines of code and re-use that in these 2 classes, would that be ok?

emteknetnz replied:
Sure, trait, extension, public static method, service class, use whatever makes the most sense given the rest of the code in the module

emteknetnz commented on tests/php/Extensions/ContentReviewCMSExtensionTest.php:
```diff
@@ -53,7 +54,7 @@ public function testGetReviewContentFormThrowsExceptionWhenObjectCannotBeReviewe
 
         // Return a DataObject without the content review extension applied
-        $mock->expects($this->once())->method('findRecord')->with(['ID' => 123])->willReturn(new Member);
+        $mock->expects($this->once())->method('findRecord')->with(['ID' => 123])->willReturn(new SiteTree);
```
Why is the return type changing here?

satrun77 replied:
The return type for `ContentReviewCMSExtension::findRecord` is `SiteTree`
https://github.com/silverstripe/silverstripe-contentreview/blob/4/src/Extensions/ContentReviewCMSExtension.php#L107

michalkleiner replied:
Does this mean the test didn't work before (and may still not work now)?

satrun77 committed:
Allow for optional can permission method for content review

emteknetnz commented on src/Extensions/SiteTreeContentReview.php:
```diff
@@ -542,6 +542,11 @@ public function canBeReviewedBy(Member $member = null)
         }
 
+        // Whether or not a user is allowed to review the content of the page.
+        if ($this->owner->hasMethod("canReviewContent") && !$this->owner->canReviewContent($member)) {
```
```suggestion
        if ($this->owner->hasMethod('canReviewContent') && !$this->owner->canReviewContent($member)) {
```
Use single quotes

jareddreyerss replied:
Fixed

emteknetnz commented on src/Traits/ReviewPermission.php:
```diff
@@ -0,0 +1,27 @@
+    protected function canUseReviewContent(DataObject $record, ?Member $user = null): bool
+    {
+        // Whether or not the user is a reviewer. User must be allowed to view the page
+        $isReviewer = $record->canView($user) &&
```
```suggestion
        $canBeReviewedByMember = $record->canView($user) &&
```
Note - depends on answer to question below

jareddreyerss replied:
removed this as we can just use a ternary on the return.

emteknetnz commented on src/Traits/ReviewPermission.php:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * Whether or not a user is allowed use content review
+     */
+    protected function canUseReviewContent(DataObject $record, ?Member $user = null): bool
```
```suggestion
    protected function isContentReviewable(DataObject $record, ?Member $member = null): bool
```
`$member` is used throughout the codebase, `$user` is not

I'm finding `canUseReviewContent` to be a strangely named because it's very similar to `canReviewContent`, as well as because it's prefixed with `can` which normally lives on $records, while here it's a utility triat.  I think renaming this to `isContentReviewable` fixes this

jareddreyerss replied:
IMO with the feedback implemented this now gets confusing with  `canBeReviewedBy` method. Because this isn't actually checking the object is reviewable its only checking the user has permission but  `canBeReviewedBy` is doing a bunch of checks that could be made more clearer by its method name.

Having said that... I did not want to tackle that in this PR 🤣 . So I have gone with your suggestion 👍

emteknetnz commented on src/Traits/ReviewPermission.php:
```diff
@@ -0,0 +1,27 @@
+        // Whether or not the user is a reviewer. User must be allowed to view the page
+        $isReviewer = $record->canView($user) &&
+            $record->hasMethod('canBeReviewedBy') &&
+            $record->canBeReviewedBy($user);
```
How come we have a `$record->canBeReviewedBy($user)` method and a `$record->canReviewContent($user)` method below?  Is that intentional? Seems like we'd only want  `$record->canReviewContent($user)` ?

jareddreyerss replied:
Agreed because canBeReviewedBy is already being called on the Review form we don't need to duplicate this check.

emteknetnz commented on src/Traits/ReviewPermission.php:
```diff
@@ -0,0 +1,27 @@
+use SilverStripe\Security\Member;
+use SilverStripe\Security\Security;
+
+trait ReviewPermission
```
```suggestion
trait PermissionChecker
```
Also remember to update the filename and references to this trait

jareddreyerss replied:
done

jareddreyerss committed:
Content Review permission logic

With the new `canReviewContent()` permission checker we only need to check the permission is set for the user and `canBeReviewedBy()` will always check if the page object is due for review by its owner.

Thus removed redundant logic in `canUseReviewContent()` and accordingly renamed the class filename for additional context.

emteknetnz reviewed (approved):
Tested locally, works well
